### ISSUE TITLE
[operation] Hazelcast 3.7: PollOperation invocation failed to complete due to operation-heartbeat-timeout

### ISSUE THREAD

**Issue by maheshreddy77** [hazelcast/hazelcast#8831]
Hi,

I am using Hazelcast 3.7.
There are no other nodes in cluster.

When i do poll with timeout, I am getting below error
{
 IQueue notifyQueue=...
 ..
 Integer segmentId = notifyQueue.poll(2, TimeUnit.MINUTES);
 return segmentId;
}

Error:

```
PollOperation invocation failed to complete due to operation-heartbeat-timeout. 
Current time: 2016-09-02 05:34:46.359. 
Total elapsed time: 121499 ms. 
Last operation heartbeat: never. 
Last operation heartbeat from member: 2016-09-02 05:34:42.350. Invocation{op=com.hazelcast.collection.impl.queue.operations.PollOperation{serviceName='hz:impl:queueService', identityHash=500295452, partitionId=51, replicaIndex=0, callId=0, invocationTime=1472819549423 (2016-09-02 05:32:29.423), waitTimeout=60000, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1472819564860, firstInvocationTime='2016-09-02 05:32:44.860', lastHeartbeatMillis=0, lastHeartbeatTime='1969-12-31 16:00:00.000', target=[172.31.142.27]:5902, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
```

Regards,
Mahesh

**Comment by jerrinot** [comment]
Hello @maheshreddy77,

many thanks for reporting. It looks like a bug in invocation system.

**Comment by pveentjer** [comment]
I tried reproducing using the following program:

```
public class Main {

    public static void main(String[] args) throws Exception {
        HazelcastInstance hz = Hazelcast.newHazelcastInstance();
        IQueue notifyQueue=hz.getQueue("foo");
        System.out.println("Waiting");
        Object result = notifyQueue.poll(2, TimeUnit.MINUTES);
        System.out.println("Ready "+result);
    }
}
```

But this works fine. Can you provide a reproducer so we can figure out what is happening?

**Comment by jerrinot** [comment]
hi @maheshreddy77: do you have any update?

**Comment by jmcshane** [comment]
I can't get a consistent reproduction of this issue, but I did want to report that this is happening regularly in one application that uses Hazelcast IQueue:

com.hazelcast.core.OperationTimeoutException: SizeOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-10-20 05:27:17.982. Total elapsed time: 120999 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2016-10-20 05:27:13.705. Invocation{op=com.hazelcast.collection.impl.queue.operations.SizeOperation{serviceName='hz:impl:queueService', identityHash=253967771, partitionId=3, replicaIndex=0, callId=0, invocationTime=1476955516983 (2016-10-20 05:25:16.983), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1476955516983, firstInvocationTime='2016-10-20 05:25:16.983', lastHeartbeatMillis=0, lastHeartbeatTime='1969-12-31 19:00:00.000', target=[HOST2]:5715, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=10, /HOST1:5715->/HOST2:46870, endpoint=[HOST2]:5715, alive=true, type=MEMBER]}

This does not begin to happen immediately after the app starts, but rather after some significant workload has been placed on the Hazelcast queuing system.

**Comment by jerrinot** [comment]
@jmcshane: what Hazelcast version is this?

**Comment by jmcshane** [comment]
Version 3.7

**Comment by jerrinot** [comment]
@jmcshane: can you try it with 3.7.2? There was actually a problem in 3.7 and it's fixed in 3.7.2: https://github.com/hazelcast/hazelcast/commit/ecbd261a317677d31da80584269e957e05b6f962

**Comment by jmcshane** [comment]
Perfect, I'll do that

**Comment by jmcshane** [comment]
@jerrinot I applied the update and it took some time for these issues to reoccur, but we are experiencing problems once again.  At this time, every operation between the two hazelcast nodes is timing out with this operation-heartbeat-timeout.  Are there some network requirements that I am missing?  Would I need to look into the VMWare setup?  I have tried pinging the IPs and telnet to the corresponding member hazelcast port and both succeed even while these operations fail.


        java.util.concurrent.ExecutionException: com.hazelcast.core.OperationTimeoutException: LockOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-10-26 01:20:55.431. Total elapsed time: 10999 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2016-10-26 01:20:51.827. Invocation{op=com.hazelcast.concurrent.lock.operations.LockOperation{serviceName='hz:impl:lockService', identityHash=481872067, partitionId=174, replicaIndex=0, callId=0, invocationTime=1477459127420 (2016-10-26 01:18:47.420), waitTimeout=50, callTimeout=10000, namespace=com.hazelcast.concurrent.lock.InternalLockNamespace@ce6cf218, threadId=115}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=10000, firstInvocationTimeMs=1477459255432, firstInvocationTime='2016-10-26 01:20:55.432', lastHeartbeatMillis=0, lastHeartbeatTime='1969-12-31 19:00:00.000', target=[10.MYHOST1]:5715, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=64, /10.MYHOST2:34860->/10.MYHOST1:5715, endpoint=[10.MYHOST1]:5715, alive=true, type=MEMBER]}
                at java.util.concurrent.FutureTask.report(FutureTask.java:122)
                at java.util.concurrent.FutureTask.get(FutureTask.java:192)
                at myco.MyClass$1.run(MyClass.java:66)
                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
                at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                at java.lang.Thread.run(Thread.java:745)
        Caused by: com.hazelcast.core.OperationTimeoutException: LockOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-10-26 01:20:55.431. Total elapsed time: 9998 ms. Last operation heartbeat: never$
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:150)
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:98)
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrow(InvocationFuture.java:74)
                at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:158)
                at com.hazelcast.concurrent.lock.LockProxySupport.tryLock(LockProxySupport.java:136)
                at com.hazelcast.concurrent.lock.LockProxySupport.tryLock(LockProxySupport.java:125)
                at com.hazelcast.concurrent.lock.LockProxy.tryLock(LockProxy.java:93)
                at myco.MyClass$2.run(MyClass.java:83)
                ... 7 common frames omitted
        01:21:06.431  INFO | hz._hzInstance_1_dev.InvocationMonitorThread |305062234 INFO  c.h.s.i.o.impl.InvocationMonitor - [10.MYHOST2]:5715 [dev] [3.7.2] Invocations:272 timeouts:1 backup-timeouts:0
        01:22:16.431 ERROR | pool-8-thread-1 |305132234 ERROR myco.MyClass
        com.hazelcast.core.OperationTimeoutException: SizeOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-10-26 01:22:16.431. Total elapsed time: 65999 ms. Last operation heartbeat: never.  Last operation heartbeat from member: 2016-10-26 01:23:14.403. Invocation{op=com.hazelcast.collection.impl.queue.operations.SizeOperation{serviceName='hz:impl:queueService', identityHash=2068134089, partitionId=268, replicaIndex=0, callId=0, invocationTime=1477459202474 (2016-10-26 01:20:02.474), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1477459336434, firstInvocationTime='2016-10-26 01:22:16.434', lastHeartbeatMillis=0, lastHeartbeatTime='1969-12-31 19:00:00.000', target=[10.MYHOST1]:5715, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=64, /10.MYHOST2:34860->/10.MYHOST1:5715, endpoint=[10.MYHOST1]:5715, alive=true, type=MEMBER]}
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:150)
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:98)
                at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrow(InvocationFuture.java:74)
                at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:158)
                at com.hazelcast.collection.impl.queue.QueueProxySupport.invokeAndGet(QueueProxySupport.java:177)
                at com.hazelcast.collection.impl.queue.QueueProxySupport.invokeAndGet(QueueProxySupport.java:170)
                at com.hazelcast.collection.impl.queue.QueueProxySupport.size(QueueProxySupport.java:104)
                at com.hazelcast.collection.impl.queue.QueueProxyImpl.size(QueueProxyImpl.java:40)
                at myco.MyClass$0(MyClass.java:75)

**Comment by jmcshane** [comment]
As a note, we had been experiencing these issues for about 8 hours consecutively, trying to debug on the network side to see if there was a problem.  We resorted to completely restarting the application and the Hazelcast cluster, and the application began functioning normally without these errors.

**Comment by jerrinot** [comment]
@jmcshane: do you have (wall-clock) time in sync on all your boxes?

**Comment by jmcshane** [comment]
@jerrinot The time in the two boxes is synced to a single source

**Comment by mufumbo** [comment]
same is happening here, with ntpdate on all servers

**Comment by pveentjer** [comment]
I is likely to be caused by:

https://github.com/hazelcast/hazelcast/issues/9251

We are working on a fix.

**Comment by mufumbo** [comment]
Not sure what's happening, it does seems like a congestion problem, but we don't have much traffic and our Map and ReplicatedMaps puts and gets are getting delayed by a huge amount. Like 50 seconds sometimes!

We had to write a complete wrapper around hazelcast to make the app to continue working. Code like:

```
 public V get(Object key) {
        long start = System.currentTimeMillis();
        V result = null;
        try {
            Future<V> future = executorService.submit(new Callable<V>() {
                @Override
                public V call() throws Exception {
                    addCount("get", 1);
                    V result = (V) getBacking().get(key);
                    return result;
                }
            });

            result = future.get(replicated ? 15 : 30, TimeUnit.MILLISECONDS);

            // TODO: make this work somehow
            if (valueClass != null && !valueClass.isInstance(result)) {
                throw new Exception("expecting " + valueClass.getSimpleName() + " but found " + result);
            }
        }
        catch (TimeoutException te) {
            log.info("timed out fetching cache " + key);
        }
        catch (Exception e) {
            log.warn("couldn't get cache " + key, e);
        }
        addTime("get", System.currentTimeMillis() - start);
        return result;
    }
```

Do you have any suggestion to debug this issue @pveentjer ?

**Comment by pveentjer** [comment]
Perhaps something else is playing. But we have a well known bug in the timed blocking operations like q.poll(2,minutes). Blocked operations are periodically aborted and retried. We also have an operation heartbeat where every running operations sends a signal to machine where the invocation is done to confirm everything is fine. 

The problem is that when retrying; the operation-heartbeat isn't reset. So when the timeout detection kicks in; it will assume the operation hasn't received a heartbeat for a long time ----> and immediately aborts the invocation with a OperationTimeoutException.

This problem can easily happen in a system that isn't do a lot of work.. it is just a matter of chance.

If you are getting 50 second latencies I would suggest adding the following:

On the member side the following parameters need to be added:
-Dhazelcast.diagnostics.enabled=true
-Dhazelcast.diagnostics.metric.level=info
-Dhazelcast.diagnostics.invocation.sample.period.seconds=30
-Dhazelcast.diagnostics.pending.invocations.period.seconds=30
-Dhazelcast.diagnostics.slowoperations.period.seconds=30

On the client side the following parameters need to be added
-Dhazelcast.diagnostics.enabled=true
-Dhazelcast.diagnostics.metric.level=info

You can use this parameter to specify the location for log file:
-Dhazelcast.diagnostics.directory=/your/log/directory

This can run in production without significant overhead. I'm always have them enabled when doing benchmarking. 

In 3.8 we added another plugin to the diagnostics that provides more detailed information about MapStore/QueueStore/RingbufferStore and Cache Store/Loader latencies. This will make it easier to pinpoint which db interaction is problematic.

**Comment by mufumbo** [comment]
Do you add those parameters to the java command runtime? It doesn't write anything anywhere (we have 3.7

**Comment by mufumbo** [comment]
I mean, if you google "hazelcast.diagnostics.enabled" there are zero pages. I believe to solve this problem we need to understand what's happening with hazelcast. It's 3 days of failures and we wrote lots of wrappers to handle hazelcast always being slow. 

I really don't understand why the architecture of Map or ReplicatedMap would EVER be synchronous for get. If we configured async-fillup="true" would it be the default behaviour to return NULL if the fillup hasn't finished? What happens now is that because the get timeout, we do a lot of PUT in sequence, so the writes never stop and it's never synched.

**Comment by pveentjer** [comment]
You can add them to the java command, but you can also use Config.setProperty.

You do not get a diagnostics log file in the current working directory? 

Else set the 
```
-Dhazelcast.diagnostics.directory=/your/log/directory
```
To a specific directory and see if the file gets created.

**Comment by pveentjer** [comment]
I would suggest creating a new ticket. Issues with map.put and replicated map are likely of a different nature than the queue.poll; where we know there is a bug in the timeout handling and already has been addressed.

**Comment by pveentjer** [comment]
@jmcshane can you try 3.7.4-SNAPSHOT? It contains the fix for OperationTimeoutException for timed blocking operations. I hope your problem will be solved. 

Map.put/replicatedMap.put are not bound blocking operations; so they are not suffering from this issue. If they do suffer from OperationTimeoutExceptions, something else must be going on. For that I would like to see the diagnostics output.

**Comment by mufumbo** [comment]
@pveentjer this option in the jvm start doesn't do anything with 3.7.2 "-Dhazelcast.diagnostics.directory"

I'm getting both replicated map and this issue. Probably they're related. 
It seems it's because of large latency in multi continent cluster.

**Comment by MattNohelty** [comment]
@pveentjer - You mentioned above that issue 9251 affects timed blocking operations like queue.poll.  Would this bug also cause operation-heartbeat-timeout in operations like map.tryLock?

**Comment by koshybobbyabraham** [comment]
@pveentjer This issue seems to have come back in 3.7.4:

Caused by: com.hazelcast.core.OperationTimeoutException: CompareAndSetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-01-17 03:07:41.734. Total elapsed time: 120698 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-01-17 03:07:37.751. Invocation{op=com.hazelcast.concurrent.atomiclong.operations.CompareAndSetOperation{serviceName='hz:impl:atomicLongService', identityHash=2002762490, partitionId=154, replicaIndex=0, callId=0, invocationTime=1484622341036 (2017-01-17 03:05:41.036), waitTimeout=-1, callTimeout=60000, name=ServiceMap_optimistic_cache_lock_next_sequence}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1484622341036, firstInvocationTime='2017-01-17 03:05:41.036', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.75.84.130]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=1, /10.75.84.132:5701->/10.75.84.130:56350, endpoint=[10.75.84.130]:5701, alive=true, type=MEMBER]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:150) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:98) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrow(InvocationFuture.java:74) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:158) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.join(AbstractInvocationFuture.java:134) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.hazelcast.concurrent.atomiclong.AtomicLongProxy.compareAndSet(AtomicLongProxy.java:81) ~[hazelcast-3.7.4.jar:3.7.4]
	at com.good.catalog.infrastructure.cluster.impl.DistributedSequenceReadWriteLockImpl.releaseReadLock(DistributedSequenceReadWriteLockImpl.java:184) ~[gd-catalog-0.54.jar:Brand_null-64]
	at com.good.catalog.infrastructure.cluster.impl.DistributedSequenceReadWriteLockImpl.access$500(DistributedSequenceReadWriteLockImpl.java:18) ~[gd-catalog-0.54.jar:Brand_null-64]
	at com.good.catalog.infrastructure.cluster.impl.DistributedSequenceReadWriteLockImpl$SequenceReadLock.unlock(DistributedSequenceReadWriteLockImpl.java:391) ~[gd-catalog-0.54.jar:Brand_null-64]
	at com.good.catalog.infrastructure.repository.hazelcast.OptimisticLockManager.releaseLock(OptimisticLockManager.java:66) ~[gd-catalog-0.54.jar:Brand_null-64]
	... 49 more

Can you suggest a fix?

**Comment by kbabr** [comment]
Some more info regarding the issue above. We have a 4 node cluster [10.75.84.130-133]. Seems like the 130 host is no longer reachable from 131-133. Here are the corresponding logs from all the 4 nodes:

10.75.84.130:
`Caused by: com.hazelcast.core.OperationTimeoutException: CompareAndSetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-01-17 03:07:41.772. Total elapsed time: 120548 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-01-17 03:07:37.752. Invocation{op=com.hazelcast.concurrent.atomiclong.operations.CompareAndSetOperation{serviceName='hz:impl:atomicLongService', identityHash=1984013857, partitionId=154, replicaIndex=0, callId=0, invocationTime=1484622341208 (2017-01-17 03:05:41.208), waitTimeout=-1, callTimeout=60000, name=ServiceMap_optimistic_cache_lock_next_sequence}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1484622341224, firstInvocationTime='2017-01-17 03:05:41.224', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.75.84.130]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
`
10.75.84.131:
`Caused by: com.hazelcast.core.OperationTimeoutException: GetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-01-17 03:07:41.382. Total elapsed time: 120885 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-01-17 03:07:37.751. Invocation{op=com.hazelcast.concurrent.atomiclong.operations.GetOperation{serviceName='hz:impl:atomicLongService', identityHash=942901448, partitionId=154, replicaIndex=0, callId=18178408, invocationTime=1484622340497 (2017-01-17 03:05:40.497), waitTimeout=-1, callTimeout=60000, name=ServiceMap_optimistic_cache_lock_next_sequence}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1484622340497, firstInvocationTime='2017-01-17 03:05:40.497', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.75.84.130]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=1, /10.75.84.131:5701->/10.75.84.130:54539, endpoint=[10.75.84.130]:5701, alive=true, type=MEMBER]}`

10.75.84.132:
`Caused by: com.hazelcast.core.OperationTimeoutException: CompareAndSetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-01-17 03:07:41.734. Total elapsed time: 120697 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-01-17 03:07:37.751. Invocation{op=com.hazelcast.concurrent.atomiclong.operations.CompareAndSetOperation{serviceName='hz:impl:atomicLongService', identityHash=935633348, partitionId=154, replicaIndex=0, callId=0, invocationTime=1484622341037 (2017-01-17 03:05:41.037), waitTimeout=-1, callTimeout=60000, name=ServiceMap_optimistic_cache_lock_next_sequence}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1484622341037, firstInvocationTime='2017-01-17 03:05:41.037', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.75.84.130]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=1, /10.75.84.132:5701->/10.75.84.130:56350, endpoint=[10.75.84.130]:5701, alive=true, type=MEMBER]}
`
10.75.84.133:
`Caused by: com.hazelcast.core.OperationTimeoutException: CompareAndSetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-01-17 03:07:41.685. Total elapsed time: 120654 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-01-17 03:07:37.750. Invocation{op=com.hazelcast.concurrent.atomiclong.operations.CompareAndSetOperation{serviceName='hz:impl:atomicLongService', identityHash=189752083, partitionId=162, replicaIndex=0, callId=0, invocationTime=1484622341033 (2017-01-17 03:05:41.033), waitTimeout=-1, callTimeout=60000, name=AppAddressMap_optimistic_cache_lock_next_sequence}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1484622341031, firstInvocationTime='2017-01-17 03:05:41.031', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.75.84.130]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=3, /10.75.84.133:57039->/10.75.84.130:5701, endpoint=[10.75.84.130]:5701, alive=true, type=MEMBER]}`

**Comment by pveentjer** [comment]
The issue with the compare and set operation is likely of a different nature than the queue.poll. There was an issue in 3.7 where blocking operations didn't send the appropriate heartbeat and therefor the caller thinks the operation isn't running ---> timeout exception. But the compare and set operation is not a blocking operation.

**Comment by jlmorton** [comment]
We're running Hazelcast on a 45 node cluster.  Recently, we've seen a problem where two nodes within the cluster cannot complete operations.

Caused by: com.hazelcast.core.OperationTimeoutException: GetOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-03-10 00:04:22.971. Total elapsed time: 119959 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-03-10 00:04:14.224. Invocation{op=com.hazelcast.concurrent.atomicreference.operations.GetOperation{serviceName='hz:impl:atomicReferenceService', identityHash=156777984, partitionId=220, replicaIndex=0, callId=0, invocationTime=1489104142701 (2017-03-10 00:02:22.701), waitTimeout=-1, callTimeout=60000, name=XYZXYZ}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1489104143012, firstInvocationTime='2017-03-10 00:02:23.012', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.X.X.51]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=64, /10.X.X.67:5701->/10.X.X.51:53171, endpoint=[10.X.X.51]:5701, alive=true, type=MEMBER]

These two members seem to be permanently in this state.  Any operations between these two nodes will fail.  These two nodes continue to participate in operations from other nodes.

Watching the communication between the nodes in Wireshark, I notice every 5 seconds a TCP ZeroWindow message from 10.X.X.51, followed by ACK from 10.X.X.67.

Looking at Netstat on 10.X.X.67, I see the send queue has been stuck on one node for several hours at 72400:

Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0  72400 10.X.X.67:5701         10.X.X.51:53171        ESTABLISHED 6572/java

Restarting the 10.X.X.51 node resolved the issue.  The node on 10.X.X.67 saw the node go down, added the member when it came back up, and after that all errors ended.

There is no network congestion between these two hosts.  However, they do share an interesting detail that they have somewhat regularly experienced major garbage collection pauses of > 60 seconds.

**Comment by bartprokop** [comment]
@jlmorton can you share what distribution / kernel version are your 10.X.X.51 and 10.X.X.67 nodes running on? Also are those server bare-metal or virtualized? Any javaagents / monitoring? Asking as seeing very similar "72400 bytes stuck" in S-Queue issue.

**Comment by jlmorton** [comment]
@bartprokop This was on CentOS Linux release 7.3.1611, Kernel 3.10.0-514.2.2.el7.x86_64.  This host was virtualized on VMWare/ESX.  No javaagents.

**Comment by fghamsary** [comment]
We have exactly the same problem on Hazelcast 3.7.4
And here is a the stack trace that we have:
`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger - com.hazelcast.core.OperationTimeoutException: PollOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-04-17 18:11:47.423. Total elapsed time: 963763009 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-04-17 18:10:45.044. Invocation{op=com.hazelcast.collection.impl.queue.operations.PollOperation{serviceName='hz:impl:queueService', identityHash=2062479137, partitionId=106, replicaIndex=0, callId=0, invocationTime=1492452707422 (2017-04-17 18:11:47.422), waitTimeout=-963600001, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1491488944414, firstInvocationTime='2017-04-06 14:29:04.414', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[127.0.0.1]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:150)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:98)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrow(InvocationFuture.java:74)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:158)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.collection.impl.queue.QueueProxySupport.invokeAndGet(QueueProxySupport.java:177)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.collection.impl.queue.QueueProxySupport.pollInternal(QueueProxySupport.java:124)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.collection.impl.queue.QueueProxyImpl.poll(QueueProxyImpl.java:88)`

`LOG:2017-04-17 20:11:47,426 INFO  [Thread1] logger -      at com.hazelcast.collection.impl.queue.QueueProxyImpl.take(QueueProxyImpl.java:82)`

Is there any workaround, or have anybody found the problem yet?

**Comment by claire-villard** [comment]
Hi,

I'm working with @fghamsary who posted the stacktrace just above.
I see that in the logs, we have `connection=null`. The target node is on the same JVM. Is it possible to have lost the node, without any other log ? 

I'm trying to reproduce the problem, but with no success for the moment.

I've dug a little in the code of Hazelcast 3.7.4, and found the method `Invocation#detectTimeout(long heartbeatTimeoutMillis)` : https://github.com/hazelcast/hazelcast/blob/v3.7.4/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java#L573.
From what I've understood, this method is where a Heartbeat timeout is detected. Several things question me in this code.

First of all : the value of the `heartbeatTimeoutMillis` parameter.
If I follow the invocation hierarchy of this method, this parameter seems to have the value of the `OPERATION_CALL_TIMEOUT_MILLIS` property. Is that normal that the call timeout is also used as Heartbeat timeout ?
For the record, this property is set to the default 60000ms in our environment.

After that, I tried to reproduce on a paper the execution of the `Invocation#detectTimeout` method, using the values written in the log.

First check : the `pendingResponse` is `VOID`. Execution continues, OK.
Second check : the `callTimeoutMillis` is `60000`, so call timeout is not disabled. Execution continues.
Third check : `op.getInvocationTime() = 1492452707422 (2017-04-17 18:11:47.422)`, so `deadlineMillis = 1492452707422 + 60000 = 1492452767422 (2017-04-17 18:12:47.422)`.

Here, I've a question : if the current time and invocation time displayed in the log are correct, why doesn't the method return at this point with the `NO_TIMEOUT__CALL_TIMEOUT_NOT_EXPIRED` code ?

If it's a normal behavior, let's proceed to the fourth and last check.

Here, `lastHeartbeatMillis = 0`, so `heartbeatExpirationTimeMillis = 1492452707422 + 60000 + 60000 = 2017-04-17 18:13:47.422`. As in the previous check, is the dates displayed in the logs are correct, I think the method should have returned here with the `NO_TIMEOUT__HEARTBEAT_TIMEOUT_NOT_EXPIRED` code ?

Could an issue with the times returned by `context.clusterClock.getClusterTime()` or `Clock.currentTimeMillis()` be possible ? 

If you have any hint, it would be greatly appreciated.

Thanks in advance,
Claire

**Comment by bartprokop** [comment]
@jlmorton [switching LRO off](https://kb.vmware.com/selfservice/microsites/search.do?language=en_US&cmd=displayKC&externalId=1027511) seems to fix the issue for me for similar kernel/hypervisor combination.

**Comment by pveentjer** [comment]
@jlmorton there is a network communication problem

```
Current time: 2017-03-10 00:04:22.971.
Total elapsed time: 119959 ms. 
Last operation heartbeat: never. 
Last operation heartbeat from member: 2017-03-10 00:04:14.224. 
```

Each member will send to each other member a period operation heartbeat; no matter if there are pending operations are not. In this case there was no operation heartbeat between the heartbeat for 8 minutes.

**Comment by jlmorton** [comment]
@pveentjer Sure, that might have been the initial cause of the problem, but it feels like there is a larger problem that Hazelcast never recovered after the initial problem.

Note that this was a 45 node cluster, and the problem was limited to operations between these two nodes.

These two nodes continued to participate in the cluster, and they continued to send operations to each other.  However, operations between these two nodes would never complete.  While there may have initially been a network issue between the two nodes, there was not a permanent one.  At the time of my investigation, the network was operating normally between the two hosts.

**Comment by saikiran939** [comment]
@pveentjer,

I see the main issue (OperationTimeout) had detour, Is this issue fixed or WIP or is there any separate issue that is opened ? 

We are using HZ 3.8 version and I see following error when executing a predicate for one of our feature.  There are only two nodes in the cluster and two nodes were discovered in members and able to communication was fine until this exception.

There are no exact reproducible steps as it happened only once in our environment and we would like to know if there is any workaround or fix that has happened. 

Please see below exception, I would also like to know if there is something that we could do from application point-of-view to overcome this exception:

```
com.hazelcast.core.OperationTimeoutException: QueryPartitionOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2017-06-16 17:03:13.994. Start time: 2017-06-16 17:01:14.992. Total elapsed time: 119002 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2017-06-16 16:58:35.649. Invocation{op=com.hazelcast.map.impl.query.QueryPartitionOperation{serviceName='hz:impl:mapService', identityHash=1038398541, partitionId=135, replicaIndex=0, callId=-2575, invocationTime=1497612673820 (2017-06-16 17:01:13.820), waitTimeout=-1, callTimeout=60000, name=com.pega.pegarules.cluster.internal.presence.PresenceAttributesHazelcastHeadNode}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1497612674992, firstInvocationTime='2017-06-16 17:01:14.992', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 05:30:00.000', target=[10.224.212.86]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=Connection[id=2, /10.224.128.85:49567->/10.224.212.86:5701, endpoint=[10.224.212.86]:5701, alive=true, type=MEMBER]}

       at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:151) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:101) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:75) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.query.MapQueryEngineImpl.addResultsOfPredicate(MapQueryEngineImpl.java:178) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.query.MapQueryEngineImpl.doRunQueryOnPartitionThreads(MapQueryEngineImpl.java:167) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.query.MapQueryEngineImpl.runQueryOnAllPartitions(MapQueryEngineImpl.java:128) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.query.MapQueryEngineImpl.execute(MapQueryEngineImpl.java:87) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.proxy.MapProxyImpl.executePredicate(MapProxyImpl.java:679) ~[hazelcast-3.8_1.jar:3.8]

       at com.hazelcast.map.impl.proxy.MapProxyImpl.values(MapProxyImpl.java:656) ~[hazelcast-3.8_1.jar:3.8]
```

**Comment by pveentjer** [comment]
@saikiran939 the OperationTimeoutException is a natural exception; it happens when the invocation has no clue what is happening to operation and assumes it must be dead.

Lets have a closer look at the message:
```
com.hazelcast.core.OperationTimeoutException: 
QueryPartitionOperation invocation failed to complete due to operation-heartbeat-timeout. 
Current time: 17:03:13.994. 
Start time: 17:01:14.992. 
Total elapsed time: 119002 ms. 
Last operation heartbeat: never. 
Last operation heartbeat from member: 16:58:35.649. 
```

Every member send to every other member periodically a listing of all operations running on behalf of the other member no matter if there is actually an operation running or not. By default this is 1/4 operation calltimeout which is 60s; so every 15 seconds we should get a heartbeat.

According to the message, the current time is 
17:03:14
and the last time of a operation-heartbeat was:
16:58:36
This means that this member has not received a mandatory operation-heartbeat for 4 minutes and 50 seconds.

Since operation-heartbeats are always send every 15 seconds as priority messages and have their own isolated thread for processing, something else is going on. Somehow on a network level no traffic is being exchanged between these members (or a huge fat packet is blocking the pipeline).

**Comment by pveentjer** [comment]
@jlmorton what kind of environments are you running on? OS... and especially if you are running on bare metal or using virtualization (please give as much detail as possible). Also give details about CPU/member (so model, number of sockets and amount of ram).

**Comment by saikiran939** [comment]
@pveentjer Thanks for the explanation. We were running on a Windows Desktop machine (not a vm or AWS variant). 

As this happened only once till now, we will keep such occurrences. Presently we are running on 3.8.1 we can make use of the 3.9 plugins you mentioned for debugging purposes.

**Comment by pveentjer** [comment]
Can you enable diagnostics

http://docs.hazelcast.org/docs/latest-development/manual/html/Management/Diagnostics/Enabling_Diagnostics_Logging.html

The key thing for me to know is what the connections are doing. THe diagnostics output file (dedicated human readable log file for HZ) contains idle time per connection. It will also show if data is pending; if there is data pending and nothing gets written.... then it smells like a problem.

**Comment by pveentjer** [comment]
@saikiran939 how many nodes do you have on that cluster on the windows machine?

**Comment by pveentjer** [comment]
@saikiran939 can you provide some more detail about your enviroment. I guess you are running on 2 different boxes since I see 2 different ip's.

10.224.128.85:49567->/10.224.212.86:5701

**Comment by pveentjer** [comment]
I'm going to close this ticket due to inactivity. Please feel free to reopen it.

**Comment by cforce** [comment]
@bartprokop Can you explain why switching off LRO fixes the issue ? Seeems like that did the trick for us too. Before i also had "Last operation heartbeat: neverLast operation heartbeat: never" in my logs ..and now  none of this. I really feel not to much comfortable that hazlecast needs such tweaks, to get performing stable. I think therefore we should reopen the issue and find out  whats the reason.

**Comment by mrumpf** [comment]
I can re-assure what @cforce said. We were able to reproduce the issue with LRO turned on under high load. The issue does not occur in low traffic scenarios (in our case below 1500 req/s). The LWN article states that LRO kicks in on 10GBit network interfaces only when the CPU is not able to deal with the high amount of network packages.
We had the parameter set with ethtool to off and the issue was gone. Yesterday one of the machines was rebooted and the LRO parameter came back to ON again.
In our high traffic phase the issue re-appeared with the one node showing response times of more than 10s (!).
The parameter seems to be dependent on the operating system. We have currently 3 different Linux versions in use RHEL 6.8, OracleLinux 7.2 and 7.3. On RHEL 6.8 LRO support was added but not enabled by default:
https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/6.8_Release_Notes/new_features_kernel.html
On RHEL 7.2 LRO has been turned on by default:
https://access.redhat.com/solutions/2168261
So a bug report without the Linux version and the LRO parameter setting does not make sense.

What is missing is a detailed explanation of what exactly is going on when the issue occurs. Maybe someone with Hazelcast and in-depth networking knowledge can step up :)

**Comment by bartprokop** [comment]
@cforce @mrumpf I do not know why switching LRO have fixed the issue. My theory is that LRO somehow interferes with kernel capacity to read data from "network interface". It was very long and non-trivial investigation to pinpoint this.  What I have noticed is that Hazelcast uses TCP connection in very specific way, actually the TCP pipes are used in unidirectional way (solely for writes or reads only, never mixture of read/write). It seems that the number of bytes send on one end usually matches the number of bytes on the other end (i.e. per IO operation). When the issue is about to happen no. of data written and read are no longer aligned between nodes (looks like bytes available/written are group into different size "packets". Then communication stops for that particular TCP pipe. Hope those observations might be somehow helpful for someone with knowledge of Hazelcast net IO internals.

**Comment by jerrinot** [comment]
@bartprokop: many thanks for the analysis you have done! 
We will have another look and knowing the issue depends on the LRO  switch is a massive help!

**Comment by cforce** [comment]
Great to hear that. In times of virtualization and clouds (PaaS)) it's uncomfortable requirement to set LRO on is layer.

**Comment by hrushig** [comment]
Our issue is too resolved by disabling LRO at the guest level. Interestingly, the cpu consumption reduced down by 10-15%.

**Comment by Batter2014** [comment]
Having same issue on PROD servers:
Cause: com.hazelcast.core.OperationTimeoutException: GetOperation invocation failed to complete due to operation-heartbeat-timeout

Hazelcast 3.8.3
Red Hat 6.4

LRO is on:
ethtool -k eth1 | grep large
Returns:
large-receive-offload: on

Server has at least 2 eth interfaces.
Why it matters? 

Found interesting blog: http://ehaselwanter.com/en/blog/2014/11/02/mtu-issue--nope-it-is-lro-with-bridge-and-bond/

Which notes:

> From the [Base Driver for the Intel® Ethernet 10 Gigabit PCI Express Family of Adapters README](http://downloadmirror.intel.com/22919/eng/README.txt:

> WARNING: The ixgbe driver compiles by default with the LRO (Large Receive Offload) feature enabled. This option offers the lowest CPU utilization for receives, but is completely incompatible with routing/ip forwarding and bridging. If enabling ip forwarding or bridging is a requirement, it is necessary to disable LRO using compile time options as noted in the LRO section later in this document. The result of not disabling LRO when combined with ip forwarding or bridging can be low throughput or even a kernel panic.

My best guess is that hazelcast from some time could act as router ("IP Forwarding"). Request comes to one eth interface and moves to another for some reason.
I'm not sure that's actually the reason.

Another point. Does anyone knows from which version issue reproduced? Most probably i will have to downgrade Hazelcast as safer option. Rather than changing server configs.

**Comment by vkandarpa** [comment]
Having the same issue in our servers also, but we have the network configuration as below

LRO is OFF on the links

We have bonding enabled and the LRO is ON on bonding. Do we need to turn off LRO on bonding also...

**Comment by Batter2014** [comment]
Just recently we stopped receiving issues related to OperationTimeoutException (without any Hazlecast/network changes) .
It's already for couple days. So i'm not sure how it will be in long run. But what was the major change on our side is removing framework which was causing 'Too Many Open Files'.
Now when I monitor app via 'lsof' i see slightly better numbers and no exceptions.

**Comment by jerrinot** [comment]
@Batter2014: many thanks for heads-up!

**Comment by cforce** [comment]
It's not sovled in hazzlecast anyway, because always any physical system, vm and container with have the defaul value o LRO on which save's a lot of CPU performance.  Hazzlecast needs to swith of this cost saver and that might be even difficlut to manage on systems you doN't 100% percent manage (onprem) but use CAAS or PAAS in the cloud.
So there is a bad work around, but thats far from solved.  Its even not explained officaily why that is necessarry,

**Comment by jerrinot** [comment]
Many thanks to all contributors to this discussion. I thought I would add a few observations I have made in recent weeks.

Apparently LRO does not play nicely with packet routing/bridging and with interface bonding. This is not specific to Hazelcast at all, it's just the way how LRO works.

See for example:
http://ehaselwanter.com/en/blog/2014/11/02/mtu-issue--nope-it-is-lro-with-bridge-and-bond/
https://github.com/moby/moby/issues/32023 

I believe bridging/routing or even bonding is routinly used in virtualized environments and you should disable LRO to be on the safe side. See [this VMWare knowledgebase](https://kb.vmware.com/s/article/1027511) article.

Also the Intel [ixgbe driver documentation](https://downloadmirror.intel.com/22919/eng/README.txt) has a good read: 
```
WARNING:  The ixgbe driver compiles by default with the LRO (Large Receive
Offload) feature enabled.  This option offers the lowest CPU utilization for
receives, but is completely incompatible with *routing/ip forwarding* and
*bridging*.  If enabling ip forwarding or bridging is a requirement, it is
necessary to disable LRO using compile time options as noted in the LRO
section later in this document.  The result of not disabling LRO when combined
with ip forwarding or bridging can be low throughput or even a kernel panic.
```

and from the same document:
  ```
Do Not Use LRO When Routing or Bridging Packets
  -----------------------------------------------
  Due to a known general compatibility issue with LRO and routing, do not use
  LRO when routing or bridging packets.
```

*Bottom Line:*
Disable LRO in virtualized environment or when you use network interface bonding. Enable it only when you absolutely have to and only after extensive testing
